internal/command/secrets/ver.go

```diff
@@ -0,0 +1,80 @@
+package secrets
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+var ErrVersNotFound = func(label string, ver Ver) error {
+	return fmt.Errorf("%s(%s): not found", label, ver)
+}
+
+// Ver is a key version. It is representable in a byte.
+// The value VerUnspec is treated as the lowest possible version
+// number, and encodes as byte(255).
+type Ver int
```

Comment:
How are you feeling about the versioning stuff?

My first thought when I saw this stuff about 254 being the max version was that it was way too low. That people might want to automate rolling keys every day and that they'd quickly run out of version numbers. That got me thinking about the utility of us managing key versioning for users.

If the user can't export the key material, they shouldn't be worried about key rotation as a matter of general hygiene. If their app was compromised, they might want to roll their keys, but do we need to facilitate that for them? If they're doings lots of encryption with a block cipher with a small nonce, they might need to roll their key periodically, but our default AEAD is going to have a large nonce, so that shouldn't be a concern.

I realize that you added the versioning stuff because I suggested it would be a nice feature, but I'm second guessing that. What are your thoughts on it? Sorry to be bringing this into question after you've written so much code to facilitate versioning, here and in the kms code...

Reply:
I'm kind of liking it so far.  The UX is pretty clean.  if you use flyctl to create "secret" with no special flags, it creates "secretv0", and when you encrypt with `/.fly/kms/secretv0/encrypt` It uses that key. If you use flyctl to create "secret" again, it creates "secretv1".  Now when you encrypt with kms, it uses secretv1, but when you use decrypt with kms, it will pick secretv0 or secretv1 based on which secret was previously used to encrypt. Its simple to use and mostly transparent.
If you don't want to use key versioning, you can use the "-n" flag when creating a secret and it will just make "secret", and overwrite it when you create it more than once.  You almost dont pay anything for the versioning feature you dont want, other than the one extra byte that gets put into the ciphertexts that you didnt need.
I agree that having a max of 256 keys might be a limitation. Though I think for the use cases (as you pointed out, keys arent disclosed, so key rotation could be a pretty rare event), it might be sufficient for most.  If we want to go bigger, it would be easy to up the size to 2 or 4 bytes, adding slight bloat to the ciphertexts.  (worst case is nacl box, which needs two key versions in ciphertext).

That said, if we want to KISS, and get rid of the version of key versioning, users who need versioning could still roll their own on top of unversioned KMS ops... but it does feel like an easy feature that does what most customers will want with little effort on our part and close to zero effort for customers.

Reply:
Cool. I'm sold on versioning then :-)

Could we do some variable-length encoding to allow for more versions without extra bloat? In the ciphertext, we could use the first two bits of the first version byte to indicate how many bytes of version information are included. 

- `00` - 1 byte - versions 0-63
- `01` - 2 bytes - versions 64-16383
- `10` - 3 bytes - versions 16384-4194303
- `11` - 4 bytes - versions 4194304-1073741823

Reply:
Rather than coding my own, I decided to go with encoding/binary's Uvarint format (based on protobufs). Its slightly less compact, but still offers a large range.

Reply:
Nice. I wasn't aware of that!